spmd/tensor/api.py

```diff
@@ -29,43 +102,72 @@ class DTensor(torch.Tensor):  # pyre-ignore[13]: pyre is bad at __new__
     @staticmethod
     def __new__(
         cls,
+        local_tensor: torch.Tensor,
         device_mesh: DeviceMesh,
         placements: Sequence[Placement],
-        size: torch.Size,
         # pyre-fixme[2]: Parameter must be annotated.
         **kwargs,
     ) -> "DTensor":
-        # new method instruct wrapper tensor and add placement spec
-        # it does not do actual distribution, __init__ should do it instead.
-        # TODO: implement __init__ for tensor constructors
-        assert isinstance(placements, list)
-        # sizes = _flatten_tensor_size(size)
-        dtype = kwargs["dtype"]
-        layout = kwargs["layout"]
-        requires_grad = kwargs["requires_grad"]
-        strides = kwargs["strides"]
-        device = kwargs.get("device", "cpu")
+        # TODO: add a docstr about tensor constructor
+        # from_local, and distribute_tensor difference.
 
+        # recover tensor shape and strides in the case of sharding
+        tensor_shape = list(local_tensor.size())
+        tensor_stride = list(local_tensor.stride())
+        for idx, placement in enumerate(placements):
+            if isinstance(placement, Shard):
+                shard_dim = placement.dim
+                local_dim_size = tensor_shape[shard_dim]
+                # recover tensor shape on the shard dim
+                tensor_shape[shard_dim] = local_dim_size * device_mesh.size(idx)
+
+                # recover tensor stride by modifying the stride that larger than
+                # the current stride on the shard_dim
+                for i in range(len(tensor_stride)):
+                    if (
+                        i < shard_dim
```

Comment:
You don't want this condition.
A larger dimension might have been transposed after the shard_dim and you still want to update it!